sycl/doc/design/IVDEPAndParallelLoopAnnotations.md

```diff
@@ -0,0 +1,338 @@
+conservative assumption that the distance is 1.
+
+In order to mark multiple arrays/pointers using ivdep on the same loop, then
+multiple instances on `ivdep(array)` or `ivdep(array,safelen)` can be used. The
+example below has the two arrays A and B marked with ivdep, but not C.
+Additionally, we indicate that the minimum dependence distance on accesses of B
+is 5 iterations.
+
+```cpp
+[[intel::ivdep(A)]]
+[[intel::ivdep(B,5)]]
+for(int i = 0; i <N; ++i) {
+   A[i] += A[idx_A[i]];
+   B[i] += B[idx_B[i]];
+   C[i] += C[idx_C[i]];
+}
+```
+
+### Marking parallel loops
+
+In contrast with `ivdep`, which only deals with memory dependences, a new
+language feature can be introduced to indicate that a loop is completely
+parallel (i.e., in addition to memory data dependences, any memory-ordering
+dependences can also be ignored). While we are not currently proposing this
+feature, we mention it for completeness and because there is an existing LLVM IR
+representation that has these semantics. If desired, it can be achieved using a
+new annotation `[[intel::parallel]]`, a new lambda-style `parallel_for` API, or
+using SYCL nested parallelism.
+
+## LLVM Metadata changes
+
+We are proposing the following new LLVM metadata:
+
+1. A new metadata, `llvm.loop.no_depends`, should be generated on loops that are
+   marked with the versions of `ivdep` that don't take a `safelen` (i.e.
+   `intel::ivdep` and `intel::ivdep(array)`). It will replace the
+   previously-generated `llvm.loop.parallel_access_indices` due to the mismatch
+   in the semantics of `llvm.loop.paralllel_accesses` and `ivdep`. The semantics
+   are that all accesses with access groups listed in this metadata have no
+   loop-carried memory data dependences.
+2. A new metadata, `llvm.loop.no_depends_safelen` should be generated on loops
+   that are marked with the versions of ivdep that take a safelen (i.e.,
+   `intel::ivdep(safelen)` and `intel::ivdep(array, safelen)`). It will replace our
+   `llvm.loop.parallel_access_indices_safelen` for the same reason described
+   above. As with `llvm.loop.no_depends`, the semantics indicate that all accesses
+   with access groups listed in this metadata have dependence distances of at
+   least safelen.
+3. The built-in `llvm.loop.parallel_accesses` metadata semantics allow the
+   compiler to not only ignore loop-carried data dependences, but also
+   loop-carried memory-ordering dependences. This metadata should not be
+   generated by the SYCL front end in conjunction with ivdep given the
+   difference in their semantics. If a parallel loop feature is added in the
+   future (as descirbed in [the previous section](#marking-parallel-loops)),
+   then this metadata would be generated from that representation.
+4. The built-in `llvm.access.group` metadata should be generated for all memory
+   operations (loads, stores, and function calls) in the body of a loop that has
```

Comment:
Should we make it clear that it's not the responsibility of the frontend to check if the call accesses memory?  It should just emit the metadata on all such function calls and leave it up to the optimizer to decide if the information is useful?